File: Dart/Structures/Queue.dart
class Node {
  Node? next;
  late int data;
  Node(int data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  Node? front;
  Node? back;
  late int size;

  Queue() {
    this.front = null;
    this.back = null;
    this.size = 0;
  }

  void enqueue(int data) {
    Node node = new Node(data);
    if (isEmpty()) {
      this.front = node;
      this.back = node;
    } else {
      this.back?.next = node;
      this.back = this.back?.next;
    }
    ++this.size;
  }

  void dequeue() {
    // handle error
    this.front = this.front?.next;
    --this.size;
  }

  void display() {
    Node? ptr = this.front;
    while (ptr?.next != null) {
      print("${ptr?.data} ");
      ptr = ptr?.next;
    }
  }

  bool isEmpty() {
    return this.front == null && this.size == 0;
  }

  int queueSize() {
    return this.size;
  }
}

void main() {
  var q = new Queue();
  for (int i = 0; i < 10; ++i) q.enqueue(i);
  q.display();
  print("dequeued");
  for (int i = 0; i < 5; ++i) q.dequeue();
  q.display();
}

File: Python/Structures/Queue.py
class Node:
    def __init__(self, data: int) -> None:
        self.data = data
        self.next = None


class Queue:
    def __init__(self) -> None:
        self.size = 0
        self.front = None
        self.back = None

    def enqueue(self, data: int) -> None:
        node = Node(data)

        if self.is_empty():
            self.front = node
            self.back = node
        else:
            self.back.next = node
            node.prev = self.back
            self.back = node

        self.size += 1

    def dequeue(self) -> None:
        if self.front is None:
            raise Exception("Error: Queue is already empty.")
        self.front = self.front.next
        self.size -= 1

    def display(self) -> None:
        ptr = self.front
        while ptr.next != None:
            print(ptr.data)
            ptr = ptr.next

    def is_empty(self) -> bool:
        return self.front == None and self.size == 0


if __name__ == "__main__":
    q = Queue()
    for i in range(0, 9):
        q.enqueue(i)

    # q.dequeue()
    q.display()
